File: LeetCode/0087. Scramble String/solution.py
"""
288 / 288 test cases passed.
Runtime: 72 ms
Memory Usage: 16.4 MB
"""
class Solution:
    @functools.lru_cache(None)
    def isScramble(self, s1: str, s2: str) -> bool:
        if s1 == s2:
            return True
        if Counter(s1) != Counter(s2):
            return False
        N = len(s1)
        for i in range(1, N):
            if self.isScramble(s1[:i], s2[:i]) and self.isScramble(s1[i:], s2[i:]):
                return True
            elif self.isScramble(s1[:i], s2[N - i:]) and self.isScramble(s1[i:], s2[:N - i]):
                return True
        return False

File: LeetCode/0087. Scramble String/solution.cpp
/**
 * 288 / 288 test cases passed.
 * Runtime: 208 ms
 * Memory Usage: 34 MB 
 */
class Solution {
public:
    unordered_map<string, bool> cache;
    bool isScramble(string s1, string s2) {
        string s = s1 + s2;
        if (cache.count(s)) return cache[s];
        if (s1 == s2) {
            cache[s] = true;
            return true;
        }
        int len = s1.size();
        for (int i = 1; i < len; ++ i) {
            string a = s1.substr(0, i), b = s1.substr(i);
            if (isScramble(a, s2.substr(0, i)) && isScramble(b, s2.substr(i)) || 
                isScramble(a, s2.substr(len - i)) && isScramble(b, s2.substr(0, len - i))) {
                    cache[s] = true;
                    return true;
            }
        }
        cache[s] = false;
        return false;
    }
};
